### DISCUSSION TITLE
Keyerror p1_current after upgrade

### DISCUSSION THREAD

**Discussion by brjhaverkamp** [addiejanssen/domoticz-solaredge-modbustcp-plugin#55] [Q&A]
Hi Addie, 

I thought this christmas break was a good idea to update my domoticz a bit. 
Your plugin was one of the things to update. I am running your plugin to great satisfaction for a long time now. 
I updated it with a simple git pull. After starting domoticz back up, 
In the log got the error: KeyError: 'p1_current' (longer snippet of my /var/log/syslog below)
Probably it is related to what I saw in your log, renaming from p1 to l1. But I don't know how to solve this.
Are there some steps to do in the database?

Regards,

Bert
Dec 21 14:49:20 joulebox domoticz[12136]: domoticz: Solar panels attic:     _plugin.onHeartbeat()
Dec 21 14:49:20 joulebox domoticz[12136]: 2022-12-21 14:49:20.133  #033[1;31mError: Solar panels attic:     _plugin.onHeartbeat()#033[0;0m
Dec 21 14:49:20 joulebox domoticz[12136]: domoticz: Solar panels attic:   File "/var/www/domoticz/plugins/domoticz-solaredge-modbustcp-plugin-backup/plugin.py", line 358, in onHeartbeat
Dec 21 14:49:20 joulebox domoticz[12136]: 2022-12-21 14:49:20.133  #033[1;31mError: Solar panels attic:   File "/var/www/domoticz/plugins/domoticz-solaredge-modbustcp-plugin-backup/plugin.py", line 358, in onHeartbeat#033[0;0m
Dec 21 14:49:20 joulebox domoticz[12136]: domoticz: Solar panels attic:     m.update(inverter_values[unit[Column.MODBUSNAME]], inverter_values[unit[Column.MODBUSSCALE]])
Dec 21 14:49:20 joulebox domoticz[12136]: 2022-12-21 14:49:20.133  #033[1;31mError: Solar panels attic:     m.update(inverter_values[unit[Column.MODBUSNAME]], inverter_values[unit[Column.MODBUSSCALE]])#033[0;0m
Dec 21 14:49:20 joulebox domoticz[12136]: domoticz: Solar panels attic: KeyError: 'p1_current'

**Comment by brjhaverkamp** [comment]
Small update. It appears the error started a bit earlier, when I upgraded domoticz to 2022.2

**Comment by brjhaverkamp** [comment]
Hi Addie, 
I meanwhile found the closed issue https://github.com/addiejanssen/domoticz-solaredge-modbustcp-plugin/issues/34
where you state " This will result in loss of historical data for these devices."
I am still coming to terms with this, as I need to throw away 2 years of data :-( 
Is there a way to edit/hack the database to save this data? and integrate it in the new device?

**Reply by addiejanssen** [reply]
Hi Bert,

The solution is to upgrade the [solaredge modbus library](https://github.com/nmakel/solaredge_modbus) as mentioned in issue #34 
My response may be outdated by now; the plugin should continue to update the existing devices, so your historical data should not get lost. Nevertheless, I would recommend to backup the database before upgrading; just to make sure...

**Reply by brjhaverkamp** [reply]
Hi Addie,

Thanks for your reply. I think I messed up significantly here. Moving from a Rpi2 to a Rpi3, updating domoticz and all plugins..
But I found that I can, as a workaround, disable the old hardware, create a new and one by one replace the devices, merging the data of the old onces with the new.

But I have  new mystery. On my freshly installed Rpi, with a new raspbian, new domoticz and new domoticz-solaredge-modbustcp-plugin install, I wasn't able to create the new hardware. In the log I got the following error:
Dec 29 10:28:41 joulebox domoticz: Solarpanels diningroom:   File "/var/www/domoticz/plugins/domoticz-solaredge-modbustcp-plugin/plugin.py", line 54, in <module>
Dec 29 10:28:41 joulebox domoticz: Solarpanels diningroom:     import solaredge_modbus
Dec 29 10:28:41 joulebox domoticz: Solarpanels diningroom:   File "/usr/local/lib/python3.9/dist-packages/solaredge_modbus/__init__.py", line 7, in <module>
Dec 29 10:28:41 joulebox domoticz: Solarpanels diningroom:     from pymodbus.client.sync import ModbusTcpClient
Dec 29 10:28:41 joulebox domoticz: Solarpanels diningroom: ModuleNotFoundError: No module named 'pymodbus.client.sync'

There seems to be a mismatch between the pymodbus and  solardege-modbus. pyython modules.
I had to replace 
#from pymodbus.client.sync import ModbusTcpClient                                                                                                                                                           
#from pymodbus.client.sync import ModbusSerialClient                                                                                                                                                        
from pymodbus.client import ModbusTcpClient
from pymodbus.client import ModbusSerialClient
(so taking out the ".client" part. 

Is this something on my side? Did I miss something? I installed both solaredge-modbux, pymodbus and (just to make sure) pymodbusTCP 
with pip3:
Requirement already satisfied: solaredge-modbus in /usr/local/lib/python3.9/dist-packages (0.7.0)
Requirement already satisfied: pymodbus>=2.4.0 in /usr/local/lib/python3.9/dist-packages (from solaredge-modbus) (3.0.2)

Regards,

Bert

**Reply by addiejanssen** [reply]
Hi Bert,

There was an issue with the solaredge_modbus library and its dependencies that got fixed in version 0.7.2.
May I suggest that you upgrade to that version?

**Reply by brjhaverkamp** [reply]
Ah, thanks for catching that. I must have missed checking for that update (I did it for the pymodbus module..) 
Indeed that was (one of) the checks.

Bert

**Reply by brjhaverkamp** [reply]
Hi Addie,

I am chasing a remaining problem in my new domoticz setup. I have regular segfaults when starting domoticz, which originate in the Solaredge-modbustcp plugin. Mind you, it could be something completely unrelated, like a corrupt SD card, but I am currenty keeping all possible causes open. So do you recognize this? Is this some setup problem (again ;-( ) with the plugin? Or should I look elsewhere?
Any advice is welcome. 

Dec 30 17:29:04 joulebox domoticz: Mosquitto Joulebox: connected to: localhost:1883
Dec 30 17:29:05 joulebox domoticz[90919]: Fatal Python error: Segmentation fault
Dec 30 17:29:05 joulebox domoticz[90919]: Current thread 0x0000007f997f9f80 (most recent call first):
Dec 30 17:29:05 joulebox domoticz[90919]:   File "<frozen importlib._bootstrap>", line 228 in _call_with_frames_removed
Dec 30 17:29:05 joulebox domoticz[90919]:   File "<frozen importlib._bootstrap_external>", line 1108 in create_module
Dec 30 17:29:05 joulebox domoticz[90919]:   File "<frozen importlib._bootstrap>", line 565 in module_from_spec
Dec 30 17:29:05 joulebox domoticz[90919]:   File "<frozen importlib._bootstrap>", line 666 in _load_unlocked
Dec 30 17:29:05 joulebox domoticz[90919]:   File "<frozen importlib._bootstrap>", line 986 in _find_and_load_unlocked
Dec 30 17:29:05 joulebox domoticz[90919]:   File "<frozen importlib._bootstrap>", line 1007 in _find_and_load
Dec 30 17:29:05 joulebox domoticz[90919]:   File "/usr/lib/python3.9/asyncio/events.py", line 786 in <module>
Dec 30 17:29:05 joulebox domoticz[90919]:   File "<frozen importlib._bootstrap>", line 228 in _call_with_frames_removed
Dec 30 17:29:05 joulebox domoticz[90919]:   File "<frozen importlib._bootstrap_external>", line 790 in exec_module
Dec 30 17:29:05 joulebox domoticz[90919]:   File "<frozen importlib._bootstrap>", line 680 in _load_unlocked
Dec 30 17:29:05 joulebox domoticz[90919]:   File "<frozen importlib._bootstrap>", line 986 in _find_and_load_unlocked
Dec 30 17:29:05 joulebox domoticz[90919]:   File "<frozen importlib._bootstrap>", line 1007 in _find_and_load
Dec 30 17:29:05 joulebox domoticz[90919]:   File "<frozen importlib._bootstrap>", line 228 in _call_with_frames_removed
Dec 30 17:29:05 joulebox domoticz[90919]:   File "<frozen importlib._bootstrap>", line 1058 in _handle_fromlist
Dec 30 17:29:05 joulebox domoticz[90919]:   File "/usr/lib/python3.9/asyncio/base_events.py", line 40 in <module>
Dec 30 17:29:05 joulebox domoticz[90919]:   File "<frozen importlib._bootstrap>", line 228 in _call_with_frames_removed
Dec 30 17:29:05 joulebox domoticz[90919]:   File "<frozen importlib._bootstrap_external>", line 790 in exec_module
Dec 30 17:29:05 joulebox domoticz[90919]:   File "<frozen importlib._bootstrap>", line 680 in _load_unlocked
Dec 30 17:29:05 joulebox domoticz[90919]:   File "<frozen importlib._bootstrap>", line 986 in _find_and_load_unlocked
Dec 30 17:29:05 joulebox domoticz[90919]:   File "<frozen importlib._bootstrap>", line 1007 in _find_and_load
Dec 30 17:29:05 joulebox domoticz[90919]:   File "/usr/lib/python3.9/asyncio/__init__.py", line 8 in <module>
Dec 30 17:29:05 joulebox domoticz[90919]:   File "<frozen importlib._bootstrap>", line 228 in _call_with_frames_removed
Dec 30 17:29:05 joulebox domoticz[90919]:   File "<frozen importlib._bootstrap_external>", line 790 in exec_module
Dec 30 17:29:05 joulebox domoticz[90919]:   File "<frozen importlib._bootstrap>", line 680 in _load_unlocked
Dec 30 17:29:05 joulebox domoticz[90919]:   File "<frozen importlib._bootstrap>", line 986 in _find_and_load_unlocked
Dec 30 17:29:05 joulebox domoticz[90919]:   File "<frozen importlib._bootstrap>", line 1007 in _find_and_load
Dec 30 17:29:05 joulebox domoticz[90919]:   File "/usr/local/lib/python3.9/dist-packages/pymodbus/client/base.py", line 4 in <module>
Dec 30 17:29:05 joulebox domoticz[90919]:   File "<frozen importlib._bootstrap>", line 228 in _call_with_frames_removed
Dec 30 17:29:05 joulebox domoticz[90919]:   File "<frozen importlib._bootstrap_external>", line 790 in exec_module
Dec 30 17:29:05 joulebox domoticz[90919]:   File "<frozen importlib._bootstrap>", line 680 in _load_unlocked
Dec 30 17:29:05 joulebox domoticz[90919]:   File "<frozen importlib._bootstrap>", line 986 in _find_and_load_unlocked
Dec 30 17:29:05 joulebox domoticz[90919]:   File "<frozen importlib._bootstrap>", line 1007 in _find_and_load
Dec 30 17:29:05 joulebox domoticz[90919]:   File "/usr/local/lib/python3.9/dist-packages/pymodbus/client/__init__.py", line 5 in <module>
Dec 30 17:29:05 joulebox domoticz[90919]:   File "<frozen importlib._bootstrap>", line 228 in _call_with_frames_removed
Dec 30 17:29:05 joulebox domoticz[90919]:   File "<frozen importlib._bootstrap_external>", line 790 in exec_module
Dec 30 17:29:05 joulebox domoticz[90919]:   File "<frozen importlib._bootstrap>", line 680 in _load_unlocked
Dec 30 17:29:05 joulebox domoticz[90919]:   File "<frozen importlib._bootstrap>", line 986 in _find_and_load_unlocked
Dec 30 17:29:05 joulebox domoticz[90919]:   File "<frozen importlib._bootstrap>", line 1007 in _find_and_load
Dec 30 17:29:05 joulebox domoticz[90919]:   File "/usr/local/lib/python3.9/dist-packages/solaredge_modbus/__init__.py", line 7 in <module>
Dec 30 17:29:05 joulebox domoticz[90919]:   File "<frozen importlib._bootstrap>", line 228 in _call_with_frames_removed
Dec 30 17:29:05 joulebox domoticz[90919]:   File "<frozen importlib._bootstrap_external>", line 790 in exec_module
Dec 30 17:29:05 joulebox domoticz[90919]:   File "<frozen importlib._bootstrap>", line 680 in _load_unlocked
Dec 30 17:29:05 joulebox domoticz[90919]:   File "<frozen importlib._bootstrap>", line 986 in _find_and_load_unlocked
Dec 30 17:29:05 joulebox domoticz[90919]:   File "<frozen importlib._bootstrap>", line 1007 in _find_and_load
Dec 30 17:29:05 joulebox domoticz[90919]:   File "/var/www/domoticz/plugins/domoticz-solaredge-modbustcp-plugin/plugin.py", line 54 in <module>
Dec 30 17:29:05 joulebox domoticz[90919]:   File "<frozen importlib._bootstrap>", line 228 in _call_with_frames_removed
Dec 30 17:29:05 joulebox domoticz[90919]:   File "<frozen importlib._bootstrap_external>", line 790 in exec_module
Dec 30 17:29:05 joulebox domoticz[90919]:   File "<frozen importlib._bootstrap>", line 680 in _load_unlocked
Dec 30 17:29:05 joulebox domoticz[90919]:   File "<frozen importlib._bootstrap>", line 986 in _find_and_load_unlocked
Dec 30 17:29:05 joulebox domoticz[90919]:   File "<frozen importlib._bootstrap>", line 1007 in _find_and_load
Dec 30 17:29:05 joulebox domoticz[90919]: domoticz: Domoticz(pid:90919, tid:90938('Solarpanels din')) received fatal signal 11 (Segmentation fault)

**Reply by addiejanssen** [reply]
Hi Bert, 

It doesn't ring a bell. Segmentation faults like this can have many causes. In the Domoticz context, most of the time it's a version conflict.
Can you list the installed software and modules?
What does `sudo pip3 list` return?

Did you install any python modules without `sudo`?
If so: can you also list the output of `pip3 list` (without sudo)?

Did you install any python module using another user account?
If so: can you list the output of `pip3 list` for all the users?

**Comment by brjhaverkamp** [comment]
Hi Addie, Thanks for your advice. I didn't know the version conflicts could be an issue. I assumed the dependencies were nicely managed by pip.
I did uninstall/reinstall all the related modules, for fear of corrupted files when I copied them from the original SD card(which I suspected of corruption.

It is strange though that domoticz starts correctly 1 out of 3-4 times. And then it stays up solidly. It is only an issue during startup.
Here is the complete module list from pip3, and all python related packages from dpkg. Is there anything you see sticking out here?

```
root@joulebox2:~# pip3 list
Package          Version
---------------- ---------
bcrypt           4.0.1
certifi          2020.6.20
cffi             1.15.1
chardet          4.0.0
colorzero        1.1
cryptography     38.0.4
distro           1.5.0
gpiozero         1.6.2
idna             2.10
numpy            1.19.5
paramiko         2.12.0
picamera2        0.3.7
pidng            4.0.9
piexif           1.1.3
Pillow           8.1.2
pip              20.3.4
pycparser        2.21
pyduro           2.0.0
pymodbus         3.0.2
PyNaCl           1.5.0
pyserial         3.5
pyserial-asyncio 0.6
python-apt       2.2.1
python-prctl     1.7
requests         2.25.1
RPi.GPIO         0.7.0
setuptools       52.0.0
simplejpeg       1.6.4
six              1.16.0
solaredge-modbus 0.7.2
spidev           3.5
ssh-import-id    5.10
toml             0.10.1
urllib3          1.26.5
v4l2-python3     0.3.1
wheel            0.34.2

root@joulebox2:~# dpkg -l | grep python
ii  libpython3-dev:arm64                 3.9.2-3                                          arm64        header files and a static library for Python (default)
ii  libpython3-stdlib:arm64              3.9.2-3                                          arm64        interactive high-level object-oriented language (default python3 version)
ii  libpython3.9:arm64                   3.9.2-1                                          arm64        Shared Python runtime library (version 3.9)
ii  libpython3.9-dev:arm64               3.9.2-1                                          arm64        Header files and a static library for Python (v3.9)
ii  libpython3.9-minimal:arm64           3.9.2-1                                          arm64        Minimal subset of the Python language (version 3.9)
ii  libpython3.9-stdlib:arm64            3.9.2-1                                          arm64        Interactive high-level object-oriented language (standard library, version 3.9)
ii  python-apt-common                    2.2.1                                            all          Python interface to libapt-pkg (locales)
ii  python-is-python3                    3.9.2-1                                          all          symlinks /usr/bin/python to python3
ii  python-pip-whl                       20.3.4-4+rpt1+deb11u1                            all          Python package installer (pip wheels)
ii  python3                              3.9.2-3                                          arm64        interactive high-level object-oriented language (default python3 version)
ii  python3-apt                          2.2.1                                            arm64        Python 3 interface to libapt-pkg
ii  python3-certifi                      2020.6.20-1                                      all          root certificates for validating SSL certs and verifying TLS hosts (python3)
ii  python3-chardet                      4.0.0-1                                          all          universal character encoding detector for Python3
ii  python3-colorzero                    1.1                                              all          Construct, convert, and manipulate colors in a Pythonic manner.
ii  python3-debconf                      1.5.77                                           all          interact with debconf from Python 3
ii  python3-dev                          3.9.2-3                                          arm64        header files and a static library for Python (default)
ii  python3-distro                       1.5.0-1                                          all          Linux OS platform information API
ii  python3-distutils                    3.9.2-1                                          all          distutils package for Python 3.x
ii  python3-gpiozero                     1.6.2-1                                          all          Simple API for controlling devices attached to a Pi's GPIO pins.
ii  python3-idna                         2.10-1                                           all          Python IDNA2008 (RFC 5891) handling (Python 3)
ii  python3-kms++                        0~git20221104~171243+adc05b6-1                   arm64        C++ library for kernel mode setting - python3 bindings
ii  python3-lib2to3                      3.9.2-1                                          all          Interactive high-level object-oriented language (lib2to3)
ii  python3-libcamera                    0~git20221130+0684c373-1                         arm64        complex camera support library (Python 3 bindings)
ii  python3-minimal                      3.9.2-3                                          arm64        minimal subset of the Python language (default python3 version)
ii  python3-numpy                        1:1.19.5-1                                       arm64        Fast array facility to the Python 3 language
ii  python3-picamera2                    0.3.7-1                                          all          Libcamera based python library
ii  python3-pidng                        4.0.9-1                                          arm64        Create Adobe DNG RAW files using Python (Python 3)
ii  python3-piexif                       1.1.3-1                                          all          Exif manipulation with pure python script (Python 3)
ii  python3-pil:arm64                    8.1.2+dfsg-0.3+deb11u1                           arm64        Python Imaging Library (Python3)
ii  python3-pip                          20.3.4-4+rpt1+deb11u1                            all          Python package installer
ii  python3-pkg-resources                52.0.0-4                                         all          Package Discovery and Resource Access using pkg_resources
ii  python3-prctl                        1.7-2+b4                                         arm64        Python interface to the prctl() syscall (Python 3)
ii  python3-requests                     2.25.1+dfsg-2                                    all          elegant and simple HTTP library for Python3, built for human beings
ii  python3-rpi.gpio                     0.7.0-0.2+b2                                     arm64        Module to control Raspberry Pi GPIO channels (Python 3)
ii  python3-setuptools                   52.0.0-4                                         all          Python3 Distutils Enhancements
ii  python3-simplejpeg                   1.6.4-1                                          arm64        Simple package for fast JPEG encoding and decoding
ii  python3-six                          1.16.0-2                                         all          Python 2 and 3 compatibility library (Python 3 interface)
ii  python3-spidev                       20200602~200721-1                                arm64        Bindings for Linux SPI access through spidev (Python 3)
ii  python3-toml                         0.10.1-1                                         all          library for Tom's Obvious, Minimal Language - Python 3.x
ii  python3-urllib3                      1.26.5-1~exp1                                    all          HTTP library with thread-safe connection pooling for Python3
ii  python3-v4l2                         0.3.1-1                                          all          Python bindings for the v4l2 userspace api
ii  python3-wheel                        0.34.2-1                                         all          built-package format for Python
ii  python3.9                            3.9.2-1                                          arm64        Interactive high-level object-oriented language (version 3.9)
ii  python3.9-dev                        3.9.2-1                                          arm64        Header files and a static library for Python (v3.9)
ii  python3.9-minimal                    3.9.2-1                                          arm64        Minimal subset of the Python language (version 3.9)
```

**Reply by brjhaverkamp** [reply]
Mmm, Not sure where the all strike-thoughs in the above message came from.??

To finish answering your questions. I have two users on the rpi: pi and bert
I did a pip3 list on both of them. The list is identical to the one for root, when I do a diff pip3.root pip3.pi  (I still have the bad habit to create a root password an work as root on my systems. I know.. I should use sudo;-) 

Regards,

Bert

**Reply by addiejanssen** [reply]
Hi @brjhaverkamp,

Unfortunately, I don't see anything strange in the lists and I have no ideas left... Looking at the information at hand, I wonder if there is a relationship with the plugin. It may _trigger_ the problem, but I don't see anything that shows that the plugin is _causing_ the problem; if that makes sense. I don't see any software conflicts in your lists either, so I wonder if we are looking in the right direction here.

If there is corruption on the file system or hiccups with the SD card it may take forever to find them. Been there; done that; learned from it: I would replace the SD card and do a clean install and start over if I would expect that that could be the case.

There's nothing else that I can think of at this moment that we could verify or test; so we are getting stuck over here. If you want to continue to investigate with the current setup, I would suggest to create a post on the Domoticz forum; there are far more people over there that like to tinker with Domoticz, the OS and software modules than over here.

Sorry for not being to help you getting this fixed.

**Comment by brjhaverkamp** [comment]
Hi Addie,
I'm indeed starting from scratch on a fresh Samsung SD card. Lets see if there is a difference.  And thanks for your help sofar. At least you confirmed I am not overlooking something simple or silly ;-) I'll keep you posted.

**Comment by brjhaverkamp** [comment]
I have a separate rpi3 with a brand new SD card and a fresh install running with the exact same symptoms. So that eliminates the SD card corruption theory, I think.
 I started a thread on the domoticz https://www.domoticz.com/forum/viewtopic.php?t=39708

**Comment by brjhaverkamp** [comment]
Hi Addie,

I might be on to something. Or at least it is an observation.
I have two sets of solar panels, both with a SolarEdge inverter. So I have two instances of the plugin running. One for the "Diningroom" and one for the "Attic".
Only if I have them both enabled, I observe the segfaults. If I have only one plugin enabled, and the other disabled, I could start domoticz 10 times without observing a crash.

Could this be a bottleneck issue or overload issue on the modbus side? or are messages confused? (I have no idea how modbus works to be honest, never dug in.)

Regards,

Bert

**Comment by brjhaverkamp** [comment]
Hi Addie,

I have been experimenting a bit tonight. I discovered that adding a configurable delay in the plugin at the start has no effect.
So it seems to start way before that. possibly even in the include solaredge-modbus line.

It has something to do with having two instances of the plugin. I assume I am a special case with two solaredge inverters.
But I tried if I could reproduce it with two instances of the plugin addressing the same inverter. 
I discovered I get the exact same segfault in that case. 
So my request is if you can try to reproduce this. Assuming you have one inverter, copy the hardware twice in domoticz and try to start domoticz a few times in a row.
/var/www/domoticz/domoticz -www 8080 -sslwww 443 -syslog -loglevel status,error

Regards,

Bert Haverkamp

**Reply by addiejanssen** [reply]
Hi Bert,

I have been trying to reproduce your error, but I have no issues after adding the inverter multiple times.
So, let's make sure that we are doing the same thing.
Which version of the plugin are you using?
What are the settings that you apply (interval, Auto Avg/Max math, Log level)?

**Reply by brjhaverkamp** [reply]
Hi Addie,

Ah, that is too bad. I was hoping I had found the smoking gun..
But now it can still be a local issue. 
Lets see.
Hardware: Raspberry Pi (Two independent ones)
SD card with latest Raspberry Pi OS Lite 64 bits
Installed domoticz with curl -L https://install.domoticz.com | bash
Installed  domoticz-solaredge-modbustcp-plugin with git clone
Installed dependencies with pip3 (the versions we already verified in an earlier mail)
I have 2 solaredge inverters in my home: an SE4K and a  SE3500H
Both have modbus over TCP enabled and are connected to my LAN by cable.
Here is a screenshot with the settings in my domoticz hardware definition.
![image](https://user-images.githubusercontent.com/3051118/212831104-ba601e63-d0d0-4df5-ad3f-4a33ba0e2df4.png)
What else is there we can check?

Bert

**Reply by addiejanssen** [reply]
Hi Bert,

I see in your screenshot that you did not select a value for the "Auto Avg/Max math" setting.
Let's make sure that that is not causing issue: please select a value from that dropdown list and update the hardware.
Do this for both inverters please.

**Reply by brjhaverkamp** [reply]
Hi Addie,

Good point. Long shot, but I put it on disabled for both hardware instances. But still two crashes out of 10 tries.
Score of the 10 tries:  (where 1 is a succesfull start and 0 is a crash, right after startup)
1110011111
I also tested an other suspicion, also a longshot, that it was still something to do with the database I moved over from the old domoticz instal. So I deleted the database and started domoticz 10x  again. Now they were all succesfull.
1111111111
Then I added the two solaredge-modbusTCP hardware instances again and tried again 10x. Now I got 3 crashes out of 10 
0101111101

So it is not the Auto Avg/max math setting nor the database:-(
From your name I estimate you are dutch. Where about in the Netherlands do you live?
I'm in Stein, Lb. Maybe I can give you physical access to the device to poke around once or so?

Regards,

Bert

**Reply by brjhaverkamp** [reply]
One more datapoint
I checked what would happen if I disconnect the two inverters from the network.
The idea being that maybe there is something in the communication at startup that kills modbus and domoticz.
First observation was that domoticz takes a long time to shut down this way:-) It is waiting about half a minute on 
Status: Solaredge attic: Stop directive received.
and an other 30 seconds on
Status: Solaredge diningroom: Stop directive received.
But unfortunately, 4 out of 10 failed, with the same segfault pattern.
0111001110

**Comment by brjhaverkamp** [comment]
Hi Addie,

I tried to compile domoticz the past few days. After figuring out my cache was not big enough that succeeded :-)
I was hoping that compiling locally would eliminate a potential ABI problem. 
But alas also a home compiled domoticz(on the same pi) shows the same crash behaviour.
I'm still looking at  https://www.domoticz.com/forum/viewtopic.php?t=37376&start=60
as a potential source of the problem though.

**Comment by brjhaverkamp** [comment]
Hi Addie,

I went digging into potential problems in asyncio in python last night. Found some interesting disucssions on open bugs with threading. But I am not sure if they are related. 
https://github.com/python/cpython/issues/80788
https://github.com/python/cpython/issues/80788#issuecomment-1276443890
In the last one Mr van Rossum pitches in.

That raised a question from my side. What hardware do you run on and what python version?

Regards,

Bert

**Comment by brjhaverkamp** [comment]
Hi Addie, 
can you please reply?  What hardware do you run on and what python version?

Regards,

Bert

**Reply by addiejanssen** [reply]
This is reported by Domoticz:
```
Version: 2022.1 (build 14471)
Build Hash: b9e5fcf0c
Compile Date: 2022-08-25 12:17:11
dzVents Version: 3.1.8
Python Version: 3.7.3 (default, Jan 22 2021, 20:04:44) [GCC 8.3.0]
```

I am running Domoticz on Docker:
`Docker version 20.10.12, build e91ed57`

The underlaying OS is:
`Ubuntu 22.04.1 LTS (GNU/Linux 5.15.0-48-generic x86_64)`

This is a virtual machine on a VMware host.
The physical box is based on:
`Intel(R) Core(TM) i7-7700 CPU @ 3.60GHz`
